google-cloud-storage/src/main/java/com/google/cloud/storage/RetryAlgorithmManager.java

```diff
@@ -0,0 +1,118 @@
+/*
+ * Copyright 2021 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *       http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.google.cloud.storage;
+
+import com.google.api.services.storage.model.Bucket;
+import com.google.api.services.storage.model.BucketAccessControl;
+import com.google.api.services.storage.model.HmacKeyMetadata;
+import com.google.api.services.storage.model.ObjectAccessControl;
+import com.google.api.services.storage.model.Policy;
+import com.google.api.services.storage.model.StorageObject;
+import com.google.cloud.ExceptionHandler;
+import com.google.cloud.storage.spi.v1.StorageRpc;
+import com.google.cloud.storage.spi.v1.StorageRpc.RewriteRequest;
+import java.io.Serializable;
+import java.util.List;
+import java.util.Map;
+
+interface RetryAlgorithmManager extends Serializable {
```

Comment:
Why not abstract instead of interface, given additional API operations may be added in the future?

Reply:
Abstract classes aren't mockable without some additional hackery necessary, and a class can only extend one class but can implement multiple interfaces. Theoretically someone might want to provide an implementation which extends their own internal class while also implementing this interface.

We can provide a Delegating implementation which can ensure not needing to re-implement the whole interface to override a single method, while keeping the public usage api against an interface.

Reply:
Naive question: It's a breaking change if a consumer chooses to implement the interface and we add an additional method which would require them to implement the method. We've hit conversation before with Storage interface and the recommendation at that time was to use class's.

Reply:
That's fair, my intent with the interface/delegate approach was to provide the interface always, with the convenience of extending a class. We don't want people to directly extend our internal implementations as that binds those implementations into the public api preventing us from making any modifications we might need to down the road.

Theoretically, we could provide default implementations in the interface now that we have java 8, but I'm not totally sure we want to do that.

Side note, this interface is currently package private and not part of the public api. If we want we can poke ahead with the interface and see how it goes before deciding to expose it publicly.